### PR TITLE
[Chore] Flakify stablecoin

### PR DESCRIPTION
**Pull request by DMozhevitin**
## Description

Problem: Due to historical reasons, `morley`-related
repos use `niv`. However, `nix` introduced native flake
feature for reproducible evaluation.

Solution: Use flakes instead of `niv` and update CI steps
to use flakes.

## Related issue(s)

Resolves #

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](https://github.com/tqtezos/stablecoin/tree/master/README.md)
    - Haddock
  - [ ] I updated [changelog](https://github.com/tqtezos/stablecoin/tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](https://github.com/tqtezos/stablecoin/tree/master/docs/code-style.md).

#### ✓ Release Checklist

- [ ] I updated the [Gas / Transaction costs](https://github.com/tqtezos/stablecoin/tree/master/README.md#gas--transaction-costs) table in the README.
- [ ] I updated the version number in `package.yaml`.
- [ ] I updated the [changelog](https://github.com/tqtezos/stablecoin/tree/master/ChangeLog.md) and moved everything
      under the "Unreleased" section to a new section for this release version.
- [ ] (After merging) I created a new entry in the [releases](https://github.com/tqtezos/stablecoin/releases) page,
      with a summary of all user-facing changes.
    * I made sure a tag was created using the format `vX.Y.Z`
    * I attached the following artifacts from [buildkite](https://buildkite.com/serokell/stablecoin)
      * `stablecoin-client`
      * `stablecoin.fa1.2.tz`
      * `stablecoin.tz`

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -7,24 +7,23 @@ env:
...
-    && cp "./nix/sources.json" "./haskell/test/resources/"
   TEZOS_CLIENT_UNSAFE_DISABLE_DISCLAIMER: "Y"
 
 .network-tests: &network-tests
   commands:
     - eval "$FETCH_CONTRACT"
     - cd haskell/
-    - export PATH=$(nix-build ../ -A stablecoin-client)/bin:$PATH
-    - nix build -L -f .. test
+    - export PATH=$(nix build ../ .#stablecoin:exe:stablecoin-client)/bin:$PATH
```
Comment (krendelhoff2) [inline .buildkite/pipeline.yml:16]:
```suggestion
    - export PATH=$(nix build ../ .#stablecoin:exe:stablecoin-client --print-out-paths)/bin:$PATH
```
Same for every such usage

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -7,24 +7,23 @@ env:
...
     # which may need to be adjusted in case of insufficient balance errors
-    - $(nix-build ../ -A run-chain-tests
+    - $(nix-build ../ -A utils.run-chain-tests
         --argstr refill-balance 1100
         --argstr node-endpoint "$$TASTY_CLEVELAND_NODE_ENDPOINT"
         --argstr step-moneybag "$$CHAIN_TEST_MONEYBAG_SECRET_KEY"
         --argstr step-moneybag-password "$$CHAIN_TEST_MONEYBAG_PASSWORD"
         --no-out-link
-        --argstr scenario './result/bin/stablecoin-test --cleveland-mode=only-network')
+        --argstr scenario './result-1/bin/stablecoin-test --cleveland-mode=only-network')
```
Comment (krendelhoff2) [inline .buildkite/pipeline.yml:26]: consider aliasing `result-1` with `--out-path` option

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+          checks = {
+            # TODO uncomment these checks and make project satisfy them.
+
+            # checks if all packages are appropriate for uploading to hackage
+            # run-cabal-check = morley-infra.utils.${system}.run-cabal-check { inherit local-packages projectSrc; };
+
+            trailing-whitespace = pkgs.build.checkTrailingWhitespace ./.;
+
+            # reuse-lint = pkgs.build.reuseLint ./.;
```
Comment (krendelhoff2) [inline flake.nix:156]: you should add `flake.lock.license` file

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -137,11 +135,11 @@ steps:
       - weeder
     commands:
       - eval "$FETCH_CONTRACT"
-      - nix-build -A stablecoin-client -o stablecoin-static
-      - nix run -f . pkgs.upx -c upx stablecoin-static/bin/stablecoin-client -o stablecoin-client
+      - nix build .#stablecoin:exe:stablecoin-client -o stablecoin-static
```
Comment (krendelhoff2) [inline .buildkite/pipeline.yml:138]: please use `-L` flag with every `nix build` invocation, except those where we use `--print-out-paths`

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (rvem) [reply .buildkite/pipeline.yml:16]: I would also provide `--no-link`

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -123,10 +120,11 @@ steps:
     commands:
       - eval "$FETCH_CONTRACT"
       - cd haskell
-      - nix-build .. -A weeder-script
+      - nix build ../  .#weeder-script
       # weeder needs .cabal file:
-      - nix run -f.. pkgs.haskellPackages.hpack -c hpack
-      - ./result
+      - nix shell ../ .#haskellPackages.hpack -c hpack
+      # TODO figure out why there are 2 results
```
Comment (rvem) [inline .buildkite/pipeline.yml:126]: Use specific `--out-link` to avoid relying on the number of `result` symlinks

File: flake.nix [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+{
+  description = "The stablecoin flake";
+
+  nixConfig.flake-registry = "https://gitlab.com/morley-framework/morley-infra/-/raw/main/flake-registry.json";
+
+  inputs.morley-infra.url = "gitlab:morley-framework/morley-infra";
+  inputs.morley.url = "gitlab:morley-framework/morley";
+
+  outputs = { self, flake-utils, haskell-nix, morley-infra, morley, ... }:
+    (flake-utils.lib.eachSystem [ "x86_64-linux" ] (system:
```
Comment (rvem) [inline flake.nix:13]: We used to provide a static binary for releases, but seems like it doesn't make a lot of sense to add such derivation to the flake currently

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -7,18 +7,17 @@ env:
...
-    && cp "./nix/sources.json" "./haskell/test/resources/"
   TEZOS_CLIENT_UNSAFE_DISABLE_DISCLAIMER: "Y"
 
 .network-tests: &network-tests
   commands:
     - eval "$FETCH_CONTRACT"
     - cd haskell/
-    - export PATH=$(nix-build ../ -A stablecoin-client)/bin:$PATH
-    - nix build -L -f .. test
+    - export PATH=$(nix build ../.#stablecoin:exe:stablecoin-client --print-out-paths --no-link)/bin:$PATH
```
Comment (rvem) [inline .buildkite/pipeline.yml:16]: A small nitpick
```suggestion
    - export PATH=$(nix build ..#stablecoin:exe:stablecoin-client --print-out-paths --no-link)/bin:$PATH
```

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -34,28 +33,25 @@ steps:
...
-  - label: check trailing whitespace
-    command: .buildkite/check-trailing-whitespace.sh
+  - label: check
     if: *not_scheduled
+    command:
+      nix flake check -L
 
   - label: xrefcheck
-    command: "nix shell -f . pkgs.xrefcheck -c xrefcheck"
+    command: "nix shell .#xrefcheck --impure -c xrefcheck"
```
Comment (rvem) [inline .buildkite/pipeline.yml:44]: `xrefcheck` was flakified, so now you can do something like:
```suggestion
    command: nix run "github:serokell/xrefcheck?rev=d1f465048eb2d9f345dbcfcc645a07440c7ebe80"
```
After that you can safely remove all mentions of `xrefcheck` from `flake.nix`

**Review by rvem — APPROVED** [review]
LGTM, I have only few minor nitpicks

File: .buildkite/pipeline.yml (hunk 7 above)
Comment (rvem) [reply .buildkite/pipeline.yml:16]: Same in the line just below

File: flake.nix [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                  # enable haddock for local packages
+                  doHaddock = true;
+              });
+
+              # disable haddock for dependencies
+              doHaddock = false;
+            }
+            {
+              packages.stablecoin.components.library = {
+                preBuild = ''
```
Comment (rvem) [inline flake.nix:69]: You can add `preHaddock` with the same contents and it should fix the necessity to do the same workaround with copying contracts for haddock build below

**Review by krendelhoff2 — APPROVED** [review]
LGTM

**Pull request merged**